frame/clients/drag_and_drop_demo.py

```diff
@@ -0,0 +1,216 @@
+        self.feedback.set_label("(nothing)")
+        self.result_callback = result_callback
+
+        self.set_label("Drop something on me!")
+        self.drag_dest_set(Gtk.DestDefaults.ALL, [], DRAG_ACTION)
+
+        self.connect("drag-data-received", self.on_drag_data_received)
+        self.connect("drag-motion", self.on_drag_motion)
+        self.connect("drag-leave", self.on_drag_leave)
+
+        if target_mode:
+            if target_mode == EXCHANGE_TYPE_TEXT:
+                self.set_text_targets()
+            if target_mode == EXCHANGE_TYPE_PIXBUF:
+                self.set_image_targets()
+
+    def buttons(self):
+        self.set_image_targets()
+        image_button = Gtk.RadioButton.new_with_label_from_widget(None, "Images")
+        image_button.connect("toggled", self.set_image_targets)
+        text_button = Gtk.RadioButton.new_with_label_from_widget(image_button, "Text")
+        text_button.connect("toggled", self.set_text_targets)
+        button_box = Gtk.Box(spacing=6)
+        button_box.pack_start(image_button, False, False, 0)
+        button_box.pack_start(text_button, False, False, 0)
+        return button_box
+
+    def set_image_targets(self, button=None):
+        targets = Gtk.TargetList.new(None)
+        targets.add_image_targets(TARGET_ENTRY_PIXBUF, True)
+        self.drag_dest_set_target_list(targets)
+
+    def set_text_targets(self, button=None):
+        targets = Gtk.TargetList.new(None)
+        targets.add_text_targets(TARGET_ENTRY_TEXT)
+        self.drag_dest_set_target_list(targets)
+
+    def on_drag_motion(self, widget, drag_context, x, y, time):
+        actions = drag_context.get_actions().value_names
+        targets = [str(target) for target in drag_context.list_targets()]
+        self.feedback.set_label("Actions: %s, offers: %s" % (actions, targets))
+        return False
+
+    def on_drag_leave(self, widget, drag_context, time):
+        self.feedback.set_label("(leave)")
+        return False
+
+    def on_drag_data_received(self, widget, drag_context, x, y, data, info, time):
+        if info == TARGET_ENTRY_TEXT:
+            self.result_callback(EXCHANGE_TYPE_TEXT)
+            text = data.get_text()
+            self.feedback.set_label("Received text: %s" % text)
+        elif info == TARGET_ENTRY_PIXBUF:
+            self.result_callback(EXCHANGE_TYPE_PIXBUF)
+            pixbuf = data.get_pixbuf()
+            width = pixbuf.get_width()
+            height = pixbuf.get_height()
+            self.feedback.set_label("Received pixbuf with width %spx and height %spx" % (width, height))
+
+def exchange_type(text):
+    if text == "text":
+        return EXCHANGE_TYPE_TEXT
+    elif text == "pixbuf":
+        return EXCHANGE_TYPE_PIXBUF
+    else:
+        raise TypeError('Unknown exchange type: "%s"' % text)
+
+if __name__ == '__main__':
+    source_mode=EXCHANGE_TYPE_NONE
+    target_mode=EXCHANGE_TYPE_NONE
+    expect=EXCHANGE_TYPE_NONE
+    try:
```

Comment:
This could all be made simpler with [argparse](https://docs.python.org/3/library/argparse.html), but your call.